core/query/sql.ml

```diff
@@ -150,157 +150,196 @@ let order_by_clause n =
-          (* using quote_field assumes tables contains table names (not nested queries) *)
-          let tables = List.map (fun (t, x) -> buffer_concat [quote t; " as "; (string_of_table_var x)]) tables
-          in string_of_select fields tables condition os
-      | Delete { del_table; del_where } ->
-          string_of_delete del_table del_where
-      | Update { upd_table; upd_fields; upd_where } ->
-          string_of_update upd_table upd_fields upd_where
-      | Insert { ins_table; ins_fields; ins_records } ->
-          string_of_insert ins_table ins_fields ins_records
-      | With (_, q, z, q') ->
-          match q' with
-          | Select (fields, tables, condition, os) ->
-              (* Inline the query *)
-              let tables = List.map (fun (t, x) -> buffer_concat [quote t; " as "; (string_of_table_var x)]) tables in
-              let buf2 = Buffer.create 0 in
-              let q2 = string_of_query buf2 quote ignore_fields q; Buffer.contents buf2 in
-              let q = buffer_concat ["("; q2; ") as "; string_of_table_var z] in
-              string_of_select fields (q::tables) condition os
-          | _ -> assert false
+  match q with
+    | UnionAll ([], _) -> Format.fprintf ppf "%a" pp_constant "select 42 as \"@unit@\" where false"
+    | UnionAll ([q], n) -> Format.fprintf ppf "%a%a" pr_q q pp_constant (order_by_clause n)
+    | UnionAll (qs, n) ->
+      let pp_sep_union ppf () = Format.fprintf ppf " union all " in
+      let pp_value ppf x = Format.fprintf ppf "(%a)" pr_q x in
+      let pp_values ppf values = Format.fprintf ppf "%a" (Format.pp_print_list ~pp_sep:pp_sep_union pp_value) values in (*MARK: can be template-list *)
+      Format.fprintf ppf "%a%a" 
+        pp_values qs 
+        pp_constant (order_by_clause n)
+    | Select (fields, [], Constant (Constant.Bool true), _os) ->
+      Format.fprintf ppf "select %a" pr_fields fields
+    | Select (fields, [], condition, _os) ->
+      Format.fprintf ppf "select * from (select %a) as %a where %a"
+        pr_fields fields
+        pp_constant (fresh_dummy_var ())
+        pr_b condition
+    | Select (fields, tables, condition, os) ->
+        (* using quote_field assumes tables contains table names (not nested queries) *)
+        let tables = List.map (fun (t, x) -> buffer_concat [quote t; " as "; (string_of_table_var x)]) tables in 
+        pr_select ppf fields tables condition os
+    | Delete { del_table; del_where } ->
+        pr_delete ppf del_table del_where
+    | Update { upd_table; upd_fields; upd_where } ->
+        pr_update ppf upd_table upd_fields upd_where
+    | Insert { ins_table; ins_fields; ins_records } ->
+        pr_insert ppf ins_table ins_fields ins_records
+    | With (_, q, z, q') ->
+        match q' with
+        | Select (fields, tables, condition, os) ->
+            (* Inline the query *)
+            let tables = List.map (fun (t, x) -> buffer_concat [quote t; " as "; (string_of_table_var x)]) tables in
```

Comment:
You can rewrite this buffer concat as something like:
`Format.asprintf "%a" pp v`.